inetsim/man/man1/inetsim.1
.\" Automatically generated by Pod::Man 4.10 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "inetsim 1"
.TH inetsim 1 "2020-05-19" "perl v5.28.1" " "
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
inetsim \- INetSim is a suite for simulating common internet services
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBinetsim\fR
[\fB\-\-config\fR <\fIconfiguration file\fR>]
[\fB\-\-version\fR]
[\fB\-\-data\-dir\fR <\fIdata directory\fR>]
[\fB\-\-log\-dir\fR <\fIlog directory\fR>]
[\fB\-\-report\-dir\fR <\fIreport directory\fR>]
[\fB\-\-bind\-address\fR <\fIip address\fR>]
[\fB\-\-max\-childs\fR <\fImaximum child processes\fR>]
[\fB\-\-user\fR <\fIuser name\fR>]
[\fB\-\-faketime\-init\-delta\fR <\fIinitial time delta\fR>]
[\fB\-\-faketime\-auto\-delay\fR <\fIdelay time\fR>]
[\fB\-\-faketime\-auto\-incr\fR <\fIstepwidth\fR>]
[\fB\-\-session\fR <\fIsession name\fR>]
[\fB\-\-pidfile\fR <\fIpid file\fR>]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBINetSim\fR simulates common internet services like
\&\fI\s-1DNS\s0\fR, \fI\s-1HTTP\s0\fR, \fI\s-1SMTP\s0\fR or \fI\s-1POP3\s0\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-\-config\fR <\fIconfiguration file\fR>" 4
.IX Item "--config <configuration file>"
The configuration file to use, default is conf/inetsim.conf in the current directory.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
Output version information.
.IP "\fB\-\-data\-dir\fR <\fIdata directory\fR>" 4
.IX Item "--data-dir <data directory>"
The data directory to use, default is data/ in the current directory.
.IP "\fB\-\-log\-dir\fR <\fIlog directory\fR>" 4
.IX Item "--log-dir <log directory>"
The log directory to use, default is log/ in the current directory.
.IP "\fB\-\-report\-dir\fR <\fIreport directory\fR>" 4
.IX Item "--report-dir <report directory>"
The report directory to use, default is report/ in the current directory.
.IP "\fB\-\-bind\-address\fR <\fI\s-1IP\s0 address\fR>" 4
.IX Item "--bind-address <IP address>"
The \s-1IP\s0 address to bind services to.
.IP "\fB\-\-max\-childs\fR <\fImaximum child processes\fR>" 4
.IX Item "--max-childs <maximum child processes>"
The maximum number of child processes (number of parallel connections) for
each service. Default is 10.
.IP "\fB\-\-user\fR <\fIusername\fR>" 4
.IX Item "--user <username>"
User to run services. Default is 'nobody'.
.IP "\fB\-\-faketime\-init\-delta\fR <\fIinitial time delta\fR>" 4
.IX Item "--faketime-init-delta <initial time delta>"
Initial number of seconds (positive or negative) relative to current
date/time for fake time used by all services. This overrides the
option 'faketime\-init\-delta' in the configuration file.
Default is 0 (use current date/time).
.IP "\fB\-\-faketime\-auto\-delay\fR <\fIdelay time\fR>" 4
.IX Item "--faketime-auto-delay <delay time>"
Number of seconds to wait before incrementing or decrementing fake time
by amount of seconds specified with \fBfaketime-auto-incr\fR.
Setting to '0' disables this option. This overrides the option
\&'faketime\-auto\-delay' in the configuration file.
Default is 0 (disabled).
.IP "\fB\-\-faketime\-auto\-incr\fR <\fIstepwidth\fR>" 4
.IX Item "--faketime-auto-incr <stepwidth>"
Number of seconds by which fake time is incremented or decremented at
regular intervals specified by \fBfaketime_auto_delay\fR. This option only
takes effect if \fBfaketime-auto-delay\fR is enabled (not set to '0').
This overrides the option 'faketime\-auto\-incr' in the configuration file.
Default is 3600.
.IP "\fB\-\-session\fR <\fIsession name\fR>" 4
.IX Item "--session <session name>"
Session name to use, default is the \s-1PID\s0 of the parent process.
.IP "\fB\-\-pidfile\fR <\fIpid file\fR>" 4
.IX Item "--pidfile <pid file>"
The pid file to use, default is /var/run/inetsim.pid.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Start the simulation with session name 'simtest5'
.PP
.Vb 1
\&        # inetsim \-\-session simtest5
.Ve
.PP
Start the simulation with a one-day-jump to the future after 60 seconds
.PP
.Vb 1
\&        # inetsim \-\-faketime\-auto\-delay 60 \-\-faketime\-auto\-incr 86400
.Ve
.PP
dito, but with jump to the past
.PP
.Vb 1
\&        # inetsim \-\-faketime\-auto\-delay 60 \-\-faketime\-auto\-incr \-86400
.Ve
.SH "REQUIRES"
.IX Header "REQUIRES"
Perl 5, Net::Server, Net::DNS, IO::Socket, IPC::Shareable, Digest::SHA,
nfqueue (optional)
.SH "AUTHOR"
.IX Header "AUTHOR"
Matthias\ Eckert <me@inetsim.org>, 
Thomas\ Hungenberg <th@inetsim.org>


inetsim/lib/INetSim.pm
# -*- perl -*-
#
# INetSim - An internet simulation framework
#
# (c)2007-2020 Matthias Eckert, Thomas Hungenberg
#
#############################################################

my $VERSION = "INetSim 1.3.2 (2020-05-19)";

package INetSim;

use strict;
use warnings;
use POSIX;
# modules to use
use INetSim::CommandLine;
use INetSim::Config;
use INetSim::Log;
use INetSim::FakeTime;
use INetSim::Chargen::TCP;
use INetSim::Chargen::UDP;
use INetSim::Daytime::TCP;
use INetSim::Daytime::UDP;
use INetSim::Discard::TCP;
use INetSim::Discard::UDP;
use INetSim::Echo::TCP;
use INetSim::Echo::UDP;
use INetSim::Quotd::TCP;
use INetSim::Quotd::UDP;
use INetSim::Time::TCP;
use INetSim::Time::UDP;
use INetSim::HTTP;
use INetSim::Ident;
use INetSim::NTP;
use INetSim::SMTP;
use INetSim::POP3;
use INetSim::DNS;
use INetSim::TFTP;
use INetSim::Report;
use INetSim::Finger;
use INetSim::Dummy::TCP;
use INetSim::Dummy::UDP;
use INetSim::FTP;
use INetSim::Syslog;
use INetSim::IRC;


#############################################################
# Local variables

my $PPID = $$;   # Parent PID
my @childs = (); # Child PIDs


#############################################################
# Child process handling
#
sub fork_services {
    my @services_to_start = INetSim::Config::getServicesToStart();
    foreach (@services_to_start) {
        my $pid = fork();
        if ($pid) {
            # we are the parent process
            push(@childs, $pid);
        }
        elsif ($pid == 0){
            # we are the child process
            if(/\Adns\z/) {
                INetSim::DNS::dns();
            }
            elsif(/\Asmtp\z/) {
                INetSim::SMTP->run;
            }
            elsif(/\Asmtps\z/) {
                INetSim::SMTP->new({ SSL => 1 })->run;
            }
            elsif(/\Apop3\z/) {
                INetSim::POP3->run;
            }
            elsif(/\Apop3s\z/) {
                INetSim::POP3->new({ SSL => 1 })->run;
            }
            elsif(/\Ahttp\z/) {
                INetSim::HTTP->run;
            }
            elsif(/\Ahttps\z/) {
                INetSim::HTTP->new({ SSL => 1 })->run;
            }
            elsif(/\Antp\z/) {
                INetSim::NTP->run;
            }
            elsif(/\Atime_tcp\z/) {
                INetSim::Time::TCP->run;
            }
            elsif(/\Atime_udp\z/) {
                INetSim::Time::UDP->run;
            }
            elsif(/\Adaytime_tcp\z/) {
                INetSim::Daytime::TCP->run;
            }
            elsif(/\Adaytime_udp\z/) {
                INetSim::Daytime::UDP->run;
            }
            elsif(/\Aident\z/) {
                INetSim::Ident->run;
            }
            elsif(/\Aecho_tcp\z/) {
                INetSim::Echo::TCP->run;
            }
            elsif(/\Aecho_udp\z/) {
                INetSim::Echo::UDP->run;
            }
            elsif(/\Adiscard_tcp\z/) {
                INetSim::Discard::TCP->run;
            }
            elsif(/\Adiscard_udp\z/) {
                INetSim::Discard::UDP->run;
            }
            elsif(/\Achargen_tcp\z/) {
                INetSim::Chargen::TCP->run;
            }
            elsif(/\Achargen_udp\z/) {
                INetSim::Chargen::UDP->run;
            }
            elsif(/\Aquotd_tcp\z/) {
                INetSim::Quotd::TCP->run;
            }
            elsif(/\Aquotd_udp\z/) {
                INetSim::Quotd::UDP->run;
            }
            elsif(/\Atftp\z/) {
                INetSim::TFTP->run;
            }
            elsif(/\Afinger\z/) {
                INetSim::Finger->run;
            }
            elsif(/\Adummy_tcp\z/) {
                INetSim::Dummy::TCP->run;
            }
            elsif(/\Adummy_udp\z/) {
                INetSim::Dummy::UDP->run;
            }
            elsif(/\Aftp\z/) {
                INetSim::FTP->run;
            }
            elsif(/\Aftps\z/) {
                INetSim::FTP->new({ SSL => 1 })->run;
            }
            elsif(/\Asyslog\z/) {
                INetSim::Syslog->run;
            }
            elsif(/\Airc\z/) {
                INetSim::IRC->run;
            }
            elsif(/\Aircs\z/) {
                INetSim::IRC->run( SSL => 1 );
            }
        }
        else {
            error_exit("Could not fork: $!", 1);
        }
    }
    sleep 1;
}


sub handle_pid {
    my $cmd = shift;
    my $pidfile = INetSim::CommandLine::getCommandLineOption("pidfile");

    $pidfile =~ /(.*)/; # evil untaint
    $pidfile = $1;
    if ($cmd eq "create") {
        if (-f $pidfile) {
            print STDOUT "PIDfile '$pidfile' exists - INetSim already running?\n";
            exit 1;
        }
        else {
            if (! open (PIDFILE, "> $pidfile")) {
                print STDOUT "Unable to open PIDfile for writing: $!\n";
                exit 1;
            }
            print PIDFILE $PPID;
            close PIDFILE;
        }
    }
    elsif ($cmd eq "remove") {
        if (-f $pidfile) {
            unlink $pidfile;
        }
        else {
            print STDOUT "Hmm, PIDfile '$pidfile' not found (but, who cares?)\n";
        }
    }
}


sub auto_faketime {
    if (INetSim::Config::getConfigParameter("Faketime_AutoDelay") > 0) {
        my $pid = fork();
        if ($pid) {
            # we are the parent process
            push(@childs, $pid);
        }
        elsif ($pid == 0){
            # we are the child process
            INetSim::FakeTime::auto_faketime();
        }
    }
}


sub redirect_packets {
    if (INetSim::Config::getConfigParameter("Redirect_Enabled")) {
        # check for linux
        if ($^O !~ /linux/i) {
            INetSim::Log::MainLog("failed! Error: Sorry, the Redirect module does not support this operating system!", "redirect");
            return 0;
        }
        # check for nfqueue library
        eval {
            eval "use nfqueue; 1" or die;
        };
        if ($@) {
            INetSim::Log::MainLog("failed! Error: Sorry, this module requires the perl nfqueue-bindings!", "redirect");
            return 0;
        }
        # check for redirect module
        eval {
               eval "use INetSim::Redirect; 1" or die;
        };
        if ($@) {
            INetSim::Log::MainLog("failed! Error: $@", "redirect");
            return 0;
        }
        my $pid = fork();
        if ($pid) {
            # we are the parent process
            push(@childs, $pid);
        }
        elsif ($pid == 0){
            # we are the child process
            INetSim::Redirect::run();
        }
    }
}


sub rest_in_peace {
    my $count = @childs;
    my $i;

    for ($i = 0; $i < $count; $i++) {
        waitpid(-1,&WNOHANG);
        if (! (kill (0, $childs[$i]))) {
            splice (@childs, $i, 1);
            $count = @childs;
            $i--;
        }
    }
}


sub wait_pids {
    wait();
    foreach (@childs){
        waitpid($_, 0);
    }
}


sub kill_pids {
    foreach (@childs){
        kill("TERM", $_);
        waitpid($_, 0);
    }
}


sub error_exit {
    my $msg = shift;
    if (! defined $msg) {
        $msg = "Unknown error";
    }
    my $exitcode = shift;
    if (! defined $exitcode) {
        $exitcode = 1;
    }
    elsif (($exitcode !~ /\A[\d]{1,3}\z/) || (int($exitcode) < 0) || (int($exitcode > 255))) {
        print STDOUT "Illegal exit code!\n";
        $exitcode = 1;
    }

    print STDOUT "Error: $msg.\n";

    kill_pids();
    wait_pids();
    handle_pid("remove");

    exit 1;
}


#############################################################
# Main
#

sub main {
    # Parse commandline options
    INetSim::CommandLine::parse_options();

    # Check command line option 'help'
    if (INetSim::CommandLine::getCommandLineOption("help")) {
        print STDOUT << "EOF";
$VERSION by Matthias Eckert & Thomas Hungenberg

Usage: $0 [options]

Available options:
  --help                         Print this help message.
  --version                      Show version information.
  --config=<filename>            Configuration file to use.
  --log-dir=<directory>          Directory logfiles are written to.
  --data-dir=<directory>         Directory containing service data.
  --report-dir=<directory>       Directory reports are written to.
  --bind-address=<IP address>    Default IP address to bind services to.
                                 Overrides configuration option 'default_bind_address'.
  --max-childs=<num>             Default maximum number of child processes per service.
                                 Overrides configuration option 'default_max_childs'.
  --user=<username>              Default user to run services.
                                 Overrides configuration option 'default_run_as_user'.
  --faketime-init-delta=<secs>   Initial faketime delta (seconds).
                                 Overrides configuration option 'faketime_init_delta'.
  --faketime-auto-delay=<secs>   Delay for auto incrementing faketime (seconds).
                                 Overrides configuration option 'faketime_auto_delay'.
  --faketime-auto-incr=<secs>    Delta for auto incrementing faketime (seconds).
                                 Overrides configuration option 'faketime_auto_increment'.
  --session=<id>                 Session id to use. Defaults to main process id.
  --pidfile=<filename>           Pid file to use. Defaults to '/var/run/inetsim.pid'.

EOF
;
        exit 0;
    }
    elsif (INetSim::CommandLine::getCommandLineOption("version")) {
        print STDOUT "$VERSION by Matthias Eckert & Thomas Hungenberg\n";
        exit 0;
    }


    # Check if we are running with root privileges (EUID 0)
    if ( $> != 0 ) {
        print STDOUT "Sorry, this program must be started as root!\n";
        exit 1;
    }

    # Check if group "inetsim" exists on system
    my $group = INetSim::Config::getConfigParameter("Default_RunAsGroup");
    $group =~ /\A(.*)\z/; # evil untaint!
    $group = $1;    
    my $gid = getgrnam($group);
    if (! defined $gid) {
        print STDOUT "No such group '$group' configured on this system!\n";
        print STDOUT "Please create group and start again. See documentation for more information.\n";
        exit 1;
    }

    print STDOUT "$VERSION by Matthias Eckert & Thomas Hungenberg\n";

    # create pidfile
    handle_pid("create");

    # Parse configuration file
    INetSim::Config::parse_config();

    # Check if there are services to start configured, else exit
    if (! scalar(INetSim::Config::getServicesToStart())) {
        INetSim::Log::MainLog("No services to start configured. Exiting.");
        handle_pid("remove");
        exit 0;
    }

    # ignore some signal handlers during startup
    local $SIG{'INT'} = 'IGNORE';
    local $SIG{'HUP'} = 'IGNORE';
    local $SIG{'TERM'} = 'IGNORE';

    INetSim::Log::MainLog("=== INetSim main process started (PID $PPID) ===");
    INetSim::Log::MainLog("Session ID:     " . INetSim::Config::getConfigParameter("SessionID"));
    INetSim::Log::MainLog("Listening on:   " . INetSim::Config::getConfigParameter("Default_BindAddress"));
    INetSim::Log::MainLog("Real Date/Time: " . strftime "%Y-%m-%d %H:%M:%S", localtime);
    INetSim::Log::MainLog("Fake Date/Time: " . (strftime "%Y-%m-%d %H:%M:%S", localtime(INetSim::FakeTime::get_faketime())). " (Delta: " . INetSim::Config::getConfigParameter("Faketime_Delta") . " seconds)");
    INetSim::Log::MainLog(" Forking services...");
    fork_services();
    auto_faketime();
    redirect_packets();

    if ($$ == $PPID) {
        $0 = 'inetsim_main';
        sleep 2;
        # reap zombies ;-)
        rest_in_peace();
        INetSim::Log::MainLog(" done.");
        INetSim::Log::MainLog("Simulation running.");
        # catch up some signalhandlers for the parent process
        local $SIG{'INT'} = sub {kill_pids();};
        local $SIG{'HUP'} = sub {kill_pids();};
        local $SIG{'TERM'} = sub {kill_pids();};
        wait_pids();
        INetSim::Log::MainLog("Simulation stopped.");

        # create report
        if (INetSim::Config::getConfigParameter("Create_Reports")) {
            INetSim::Report::GenReport();
        }

        INetSim::Log::MainLog("=== INetSim main process stopped (PID $PPID) ===");
        INetSim::Log::MainLog(".");
    }

    # delete pidfile
    handle_pid("remove");
    exit 0;
}


1;
#
